### PR TITLE
Speed up event processing when it's lagging behind 

### PR DESCRIPTION
**Pull request by mpfz0r**
Introduce global CatchUpWindowSize setting, defaulting to 1 hour.
This allows event processors that are catching up on old data
to catch up using a bigger window than the regular processingWindowSize.
This can happen if the server was shut down for a significant amount of time.

 - Add frontend for catch up window size

 - Wrap aggregation in time histogram aggregations so they work
   on larger chunks than the configured processingWindowSize

### CODE REVIEW

**Comment by bernd** [comment]
I fixed a conflict.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.graylog.plugins.views.search.elasticsearch.searchtypes.pivot.ESPivotBucketSpecHandler;
+import org.graylog.plugins.views.search.searchtypes.pivot.Pivot;
+import org.graylog.plugins.views.search.searchtypes.pivot.buckets.DateRangeBucket;
+import org.joda.time.base.AbstractDateTime;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class ESDateRangeHandler extends ESPivotBucketSpecHandler<DateRangeBucket, DateRangeAggregation> {
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java:36]: We have some basic tests for the other bucket handler. How about creating one for this handler, too? :slightly_smiling_face:

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    public Stream<Bucket> doHandleResult(Pivot pivot,
+                                         DateRangeBucket dateRangeBucket,
+                                         SearchResult searchResult,
+                                         DateRangeAggregation rangeAggregation,
+                                         ESPivot searchTypeHandler,
+                                         ESGeneratedQueryContext esGeneratedQueryContext) {
+        // There is only one key per Bucket.
+        // We have to decide between from() or to()
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java]: OK, done :+1:

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java (hunk 1 above)
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java:36]: I'm not really sure that's worth the trouble, what is it supposed to test? There's barely any logic in there.
Besides, there is no test for `ESValuesHandler`  :-P

Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/pivot/buckets/ESDateRangeHandler.java:36]: Okay, fair enough. :smile:

File: graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorExecutionJob.java [hunk 3]
```diff
@@ -122,8 +126,21 @@ public JobTriggerUpdate execute(JobExecutionContext ctx) throws JobExecutionExce
...
+            // If the event processor is catching up on old data (e.g. the server was shut down for a significant time),
+            // we can switch to a bigger scheduling window: `processingCatchUpWindowSize`.
+            // If enabled, it will ignore the processingHopSize and will schedule a tumbling window timerange of `processingCatchUpWindowSize` chunks.
+            final long catchUpSize = configurationProvider.get().eventCatchupWindow();
+            if (catchUpSize > 0 && catchUpSize > config.processingWindowSize() && to.plus(catchUpSize).isBefore(now)) {
+                final long chunkCount = catchUpSize / config.processingWindowSize();
+
+                // Align to multiples of the processingWindowSize
+                nextTo = to.plus(config.processingWindowSize() * chunkCount);
+                LOG.debug("eventproc <{}> is catching up on old data. Combining {} search windows with catchUpWindowSize={}ms: from={} to={}",
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorExecutionJob.java:141]: @bernd I'm wondering if we should make this `LOG.info` what do you think?

Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorExecutionJob.java]: @mpfz0r Sounds good to me! But please replace `eventproc` with `Event processor` if we show it to users. :slightly_smiling_face:

Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorExecutionJob.java]: Cool. Did that already :-)

**Pull request merged**